Made portable creation through the installer or tools menu more flexible, by allowing system variables in paths

XLTechie opened the pull request:
### Link to issue number:

Fixes #14680 

### Summary of the issue:

While it is possible to use system variables (%appdata%, %userprofile%, %temp%, etc.) in the create portable dialog's "browse" sub-dialog, it was not possible to use them in the Create Portable dialog itself. Therefore, when using the installer or the tools menu portable option, one either had to enter full paths, or use the browse dialog to utilize system variables to minimize typing.
For people who create many portables, that can be unnecessary overhead.

### Description of user facing changes

It is now possible to use system variables (such as %tmp% or %homepath%) in the path specification while creating portable copies of NVDA.
Additionally, it is no longer necessary to include a drive letter when entering the absolute path to a portable target.

### Description of development approach

In `gui.installerGUI`, there is a method of `PortableCreaterDialog` class, called `onCreatePortable`: The path, as entered in the dialog (`self.portableDirectoryEdit.Value`), was used in this method three times.

- Created a local variable, `expandedPortableDirectory`, containing that value, processed through `os.path.expandvars()`.
- Used that local value throughout the remainder of the method.
- Added a note to the error which appears if the user doesn't specify an absolute path, stating that system variables may be used.
- Rephrased the messages asking for the path, and removed the mention of the required drive letter, per NV Access review comments.
- Replaced a % substitution with a .format() string formatter, in an error dialog.
- Edited, reformatted, and added missing translator comments.
- Miscellaneous linting.
Additional:
- In gui.installerGUI.doCreatePortable(): Linting, spelling, added translation comments, replaced substitutions with calls to format

#### A note regarding no longer checking for drive letter in entered absolute path:

While creating a portable copy, we ask the user for an absolute path. In the os.path module's opinion, a path can be absolute without having a drive letter. Therefore, we previously checked and threw an error if the drive letter was missing.
When @seanbudd suggested removing the drive letter notation from absolute path messages, testing without entering one revealed errors that occurred if there was no drive letter, even though it shouldn't technically matter.

Instead, since os.path.isabs has already determined that the path is absolute, we can just add the drive letter if it's missing.
We can do that by calling os.path.abspath, which absolutizes the (probably mostly) already absolute path, by adding a drive letter if it was missing, and any other path components necessary to have a fully qualified path.

### Testing strategy:

* Created portables using variables in their paths, from both the installer and the tools menu.
* Tested that badly-named variables cause the same error as invalid paths.
* Absolute paths with no drive letter work as a user would expect.

### Known issues with pull request:
None
### Change log entries:
New features
It is now possible to use system variables (such as %temp% or %homepath%) in the path specification while creating portable copies of NVDA.

Bug fixes
`gui.installerGUI.PortableCreaterDialog.onCreatePortable` now properly checks for system variables used in the entered paths of potential portable copies, instead of automatically showing an invalid path error when system variables are used.

Changes
Portable copy creation no longer requires that a drive letter be entered as part of the absolute path.

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

Commit:
Made portable creation through the installer or tools menu more flexible, by allowing system variables in paths. (#14681)

In the method PortableCreaterDialog.onCreatePortable:
The path, as entered in the dialog (`self.portableDirectoryEdit.Value`), was used in this method three times.
- Created a local variable, `expandedPortableDirectory`, containing that value, processed through `os.path.expandvars()`.
- Used that local value throughout the remainder of the method.
- Removed the mention of drive letter in absolute path messages, per NV Access suggestion.
- Added a note to the error which appears if the user doesn't specify an absolute path, stating that system variables may be used, and giving examples.
- Reformatted a translation comment for the title of that error dialog.
- Replaced a % substitution with a .format() string formatter, in an error dialog.
- Miscellaneous linting.
- Added a few missing translator comments.

seanbudd commented on source/gui/installerGui.py:
```diff
@@ -404,34 +404,46 @@ def __init__(self, parent):
+				_(
+					# Translators: The message displayed when the user has not specified an absolute destination directory
+					# in the Create Portable NVDA dialog.
+					"Please specify an absolute path (including drive letter)  in which to create the "
```
```suggestion
					"Please specify the absolute path where the portable copy should be created."
```

seanbudd commented on source/gui/installerGui.py:
```diff
@@ -404,34 +404,46 @@ def __init__(self, parent):
+				_(
+					# Translators: The message displayed when the user has not specified an absolute destination directory
+					# in the Create Portable NVDA dialog.
+					"Please specify an absolute path (including drive letter)  in which to create the "
+					"portable copy. It may include system variables (%temp%, %userprofile%, etc.)."
```
I think this could be simplified.
I think if someone is aware of the meaning of "absolute path" they are aware that it includes the drive letter.
It also seems like the logic below doesn't require a drive letter.

```suggestion
					"Please specify the absolute path where the portable copy should be created. "
					"It may include system variables (%temp%, %userprofile%, etc.)."
```

seanbudd commented on source/gui/installerGui.py:
```diff
@@ -404,34 +404,46 @@ def __init__(self, parent):
 			return
 		self.Hide()
-		doCreatePortable(self.portableDirectoryEdit.Value,self.copyUserConfigCheckbox.Value,False,self.startAfterCreateCheckbox.Value)
+		doCreatePortable(
+			expandedPortableDirectory, self.copyUserConfigCheckbox.Value,
+			False, self.startAfterCreateCheckbox.Value
+		)
```
This should be split with a newline for each argument

```suggestion
		doCreatePortable(
			expandedPortableDirectory,
			self.copyUserConfigCheckbox.Value,
			False,
			self.startAfterCreateCheckbox.Value
		)
```

seanbudd reviewed (commented):
Hi @XLTechie  - generally looks good to me, added some notes.

XLTechie commented:
@seanbudd Think I have covered your wishes here. Note that the drive letter is apparently required, per line 431+. I still removed it from the message as you suggested, as I agree it seems redundant to mention given "absolute path".

seanbudd commented on source/gui/installerGui.py:
```diff
@@ -404,34 +404,48 @@ def __init__(self, parent):
 			return
-		drv=os.path.splitdrive(self.portableDirectoryEdit.Value)[0]
+		drv = os.path.splitdrive(expandedPortableDirectory)[0]
 		if drv and not os.path.isdir(drv):
```
if drv is false-y than this does not trigger, meaning that if there is no drive this continues

seanbudd replied:
I think handling the falsey case similar to an invalid drive should fix this.

Commit:
gui.installerGUI.doCreatePortable(): Linting, spelling, added translation comments, replaced substitutions with calls to format

Commit:
Stop trying to check for drive letter in entered absolute path.

When creating a portable copy, we ask the user for an absolute path.
Since technically a path can be absolute without having a drive letter, we previously checked and threw an error
if the drive letter was missing, even though Windows doesn't require one to be absolute about the path.
Instead, since os.path.isabs has already determined that the path is absolute, all that remains is to
add the drive letter if it's missing.
We can do that by calling os.path.abspath, which absolutizes the (probably mostly) already absolute path,
by adding a drive letter if it was missing, and any other path components necessary to have a pedantically correct path.

XLTechie commented:
@seanbudd please see revised development approach section of PR comment to explain what I did here and see if this method meets with your approval.

Commit:
Added a blank line (linting)

seanbudd commented on source/gui/installerGui.py:
```diff
@@ -404,69 +404,90 @@ def __init__(self, parent):
+				wx.OK | wx.ICON_ERROR
+			)
 			return
+		# We used to separate out the drive letter, check it, and present an error if it was not
+		# its own directory, as a secondary check of path absoluteness.
+		# But if os.path.isabs is convinced, we should just add the drive letter if missing, and carry on. (#14681)
+		expandedPortableDirectory = os.path.abspath(expandedPortableDirectory)
```
does abspath add the missing drive letter? what is the default? C:? the directory the installer is being run from?

I think explaining the old behaviour here might be misleading or confusing, I'd rather document the current behaviour more clearly.

```suggestion
		# isabs determines if the path is absolute, with or without a drive letter.
		# abspath resolves an absolute path, adding a drive letter (which?) if necessary. (#14681)
		expandedPortableDirectory = os.path.abspath(expandedPortableDirectory)
```

seanbudd reviewed (commented):
Approach generally LGTM, just some feedback on a comment

XLTechie committed:
Incomplete review suggestion

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

Commit:
Explain where the os.path.abspath drive letter comes from.

XLTechie commented:
@seanbudd, how's that?

It's not exactly clear what the drive will be in all possible cases, but it's 
whatever Windows considers the current drive. Which is probably (always?) the 
drive NVDA is running on.

Short of going back to mandating that a drive be chosen, I'm not sure if we can 
do any better according to the os.path docs.

We can't even confirm that the first component of the path exists, because a 
user might be using something like "d:\nvda", which isn't yet created.

seanbudd committed:
Merge remote-tracking branch 'origin/master' into fix14680

seanbudd committed:
update changes

seanbudd committed:
update changes

XLTechie commented:
Thanks @seanbudd.
